File: ruby/lib/mathlib.rb
module Mathlib
	def Factorial(n)
		if n == 1
			return 1
		else
			return n * Factorial(n - 1)
		end
	end
	
	def Ack(m, n)
		if m == 0
			return n+1
		elsif ((m > 0) and (n == 0))
			return Ack(m-1, 1)
		elsif ((m > 0) and (n > 0))
			return Ack(m-1, Ack(m, n-1))	
		end
	end
	
	def GCD(a, b)
		if b == 0
	    		return a
	    	end
		return GCD(b, a % b)
	end
	
	def LCM(a, b)
		return (a / GCD(a, b)) * b
	end
	
	def Fibonacci(n)
		if n <= 1
			return n
		end
		return Fibonacci(n - 1) + Fibonacci(n - 2)
	end
	
	def Derivative(x, n)
		return n * x ** (n - 1)
	end
end

File: lua/mathlib.lua
local mathlib = {}

function mathlib.summation(s,e)
  if s > e then
    return 0
  else
    return s + mathlib.summation(s+1,e)
  end
end

function mathlib.factorial(n) 
	if n == 1 then
		return 1
	else
		return n * mathlib.factorial(n - 1)
	end
end

function mathlib.ack(m, n)
    if m == 0 then
    	return n + 1
    elseif ((m > 0) and (n == 0)) then
    	return mathlib.ack(m - 1, 1)
    elseif ((m > 0) and (n > 0)) then
    	return mathlib.ack(m - 1, mathlib.ack(m, n - 1))
    else
    	return 0
    end
end

function mathlib.GCD(a, b)
	if b == 0 then
		return a
	end
	return mathlib.GCD(b, a % b)
end

function mathlib.LCM(a, b)
	return (a / mathlib.GCD(a, b)) * b
end

function mathlib.fibonacci(n)
	if n <= 1 then
		return n
	end
	return mathlib.fibonacci(n - 1) + mathlib.fibonacci(n - 2)
end

function mathlib.derivative(x, n)
	return n * x ^ (n - 1)
end

return mathlib
